File: src/python/flesch.py
#!/usr/bin/python3
import sys

#################################### 
class flesch:
  def __init__(self):               #Building a flesch-readability object with attributes
    self.syllableCount=0
    self.wordCount=0
    self.sentenceCount=0
    self.diffWordCount=0
#################################### 

####################################     
def main(argv):
  index=countWords(sys.argv[0])
  a = (index.syllableCount)/(index.wordCount)   #
  b = (index.wordCount)/(index.sentenceCount)   #Variables used for respective equations
  aD = (index.diffWordCount)/(index.wordCount)  #
  print("Difficult Word: "+str(index.diffWordCount))
  
  readIndex = 206.835-(a*84.6)-(b*1.015)        #Flesch-Readability Index Variable
  gradeIndex = (a*11.8)+(b*0.39)-15.59          #Flesch Grade Level Index
  dalePercent = (aD * 100)                      #Dale-Chall Percentage
  
  daleIndex =(dalePercent*0.1579)+(b*0.0496)    #Dale-Chall Raw Score
  if(dalePercent > 5):
    daleIndex += 3.6365
  print("Sentence Count: "+str(index.sentenceCount))
  print("Word Count: "+str(index.wordCount))
  print("Syllable Count: "+str(index.syllableCount))
  print("\nReadability and Grade Index")
  print("-----------------------------------")
  print("Flesch Readability Level: "+str(round(readIndex)))
  print("Flesch Grade Level Index: "+str(round(gradeIndex, 1)))
  print("Dale-Chall Readability Level: "+str(round(daleIndex, 1)))
#################################### 

#################################### 
def countWords(args):                         #Probably should rename this to buildFleschObject()
  dalechall=buildHashSet()                    #Unordered set of Dale-Chall words
  index=flesch();
  with open(sys.argv[1], 'r') as inFile:      #
    for line in inFile:                       #Reading in file and splitting based on white space
      for word in line.split():               #
        if(isWord(word)):
          i=0
          for i in word:
            if(isSentence(i)):
              index.sentenceCount+=1
          word = checkWord(word)                   
          if(isDifficultWord(dalechall,word)):          #
            index.diffWordCount+=1                      #
          index.wordCount+=1                            #
          index.syllableCount+=countSyllables(word)     #                                          #
  return index                                          #returning built Flesch Object
####################################       

####################################
def countSyllables(word):
  count = 0
  hasSyllable=False              #boolean to check if word has syllables
  state = 'c'                    #Default state is 'c' for cosonant
 
  end = word[-1:]
    
  for i in word:                #Simple state machine inspired by a Stack Overflow answer: https://stackoverflow.com/a/52649782
                                
    if(state=='v'):             #If the previous state of the machine was a vowel
      
      if(not isVowel(i)):   #and if the current state of the machine is not a vowel, increment. This accounts for consecutive vowels.
        hasSyllable=True
        count+=1
          
    state='v' if isVowel(i) else 'c' #Switching states depending on current character
      
  if((state == 'v' and end!='e' and hasSyllable)):           #Special case for when the word ends in 'e'                         
    count+=1;
  elif(state=='v' and (not hasSyllable)):             #Special case for when the word ends in 'e' but has no other syllables
    count+=1;
  return count;
####################################

#################################### 
def isVowel(i):
  i=i.lower()
  if(i=='a'or i=='e' or i=='i' or i=='o' or i=='u' or i=='y'):
    return True;
  return False;
####################################

#################################### 
def isSentence(i):
  if(i=='.' or i==':' or i==';' or i=='!' or i=='?'):
    return True;
  return False;
####################################

#################################### 
def isWord(word):
  i=0
  
  for i in word:                  #Iterates through word, if the word has at least one letter then it is considered a word. 
    if(i.isalpha()):
      return True
      
  return False
####################################

#################################### 
def isDifficultWord(dalechall, word):
  word=word.lower()
  return(not(word in dalechall))        #Returning the negation of the value, i.e. a word is a difficult word if it is NOT in the Dale-Chall list
#################################### 

####################################   
def buildHashSet():
  dalechall = set()
  with open("/pub/pounds/CSC330/dalechall/wordlist1995.txt") as inFile:
    for line in inFile:
      line=line.rstrip('\n')
      dalechall.add(line)
  return dalechall
#################################### 
def checkWord(word):
  if(word[0]=='['):
    word = word[1:]
  if(word[-1]==']' or word[-1]==',' or isSentence(word[-1])):
    word = word[:-1]
  return word
#################################### 
if __name__=='__main__':
  main(sys.argv[1:])

File: src/perl/flesch.pl
#!/usr/bin/perl
use strict;
use warnings;

package flesch;
sub new {
  my $class = shift;
  my $self = {
    syllableCount=>shift,
    wordCount=>shift,
    sentenceCount=> shift,
    diffWordCount=>shift,
    };
    bless $self, $class;
    return $self;
}

sub buildFleschObject {
## Grab the name of the file from the command line, exit if no name given
 my @dalechall = buildHashSet();
 #print "@dalechall"; 
 my $fleschIndex = new flesch(0,0,0,0);
 
 my $filename = $_[0] or die "Need to get file name on the  command line\n";
 
 ## Use the filename 
 open(DATA, "<$filename") or die "Couldn't open file $filename, $!"; 
 
 ##the next line puts all the lines from the text file into an array called @all_lines
 my @all_lines = <DATA>; 
 
 ##Now take each line and break it into tokens based on spaces and print the token
 foreach my $line (@all_lines) {
    my @words = split(' ', $line);
       foreach my $word (@words) {
         #print "$word\n";
            if(isWord($word))
            {
              if(isDifficultWord(@dalechall, $word))
              {
                $fleschIndex->{diffWordCount}++;
              }
              #print "Word to be checked: $word\n";
              $fleschIndex->{wordCount}++;
              $fleschIndex->{syllableCount}+=countSyllables($word);
              foreach my $char (split //, $word)
              {
                if(isSentence($char))
                {
                  $fleschIndex->{sentenceCount}++;
                }
              }
            }
       }
 }
 return $fleschIndex;
}


sub countSyllables {
  my $count = 0;
  my $hasSyllable = 0;
  my $state = "c";
  
  my $word = $_[0];
  my $end = substr($word, -1);
  #print "$word";
  #print "$end";
  if(isSentence($end))
  {
    chomp $word;
    $end = substr($word, -1);
  }
  foreach my $char (split //, $word)
  {
    if($state eq "v")
    {
      if(not(isVowel($char)))
      {
        $hasSyllable = 1;
        $count++;
      }
    }
    $state = isVowel($char)?"v":"c";
  }
  if(($state eq "v") and (not ($end eq "e")) and ($hasSyllable))
  {
    $count++;
  }
  elsif(($state eq "v") and (not($hasSyllable)))
  {
    $count++;
  }
  return ($count);
}

sub isVowel {
  my $char = $_[0];
  $char = lc($char);
  my @vowels = ("a","e","i","o","u","y");
  my %hashmap = map {$_ => 1 } @vowels;
  return (exists($hashmap{$char}));
}

sub isSentence {
  my $char = $_[0];
  ##print "$char";
  my @punctuations = (".",":",";","!","?");
  my %hashmap = map {$_ => 1 } @punctuations;
  return (exists($hashmap{$char}));
}

sub isWord {
  my $word = $_[0];
  $word = lc($word);
  foreach my $char (split //, $word)
  {
  if($char =~ /^[a-z]+$/i)
  {
    return 1;
  } 
  }
  
  return 0;
}

sub isDifficultWord
{
  my @dalechall = $_[0];
  my %hashmap = map {$_ => 1 } @dalechall;
  my $word = $_[1];
  my $end = substr($word, -1);
  print "Checking: $word\n";
  #print "$end";
  if((isSentence($end)) or ($end eq ","))
  {
    chomp $word;
    $end = substr($word, -1);
  }
  return(not(exists($hashmap{$word})));
}

sub buildHashSet
{
  my @dalechall;
  my $filename = "/pub/pounds/CSC330/dalechall/wordlist1995.txt";
  open(DATA, "<$filename") or die "Couldn't open file $filename, $!"; 
  my @all_lines = <DATA>; 
  foreach my $line (@all_lines) {
    my @words = split(' ', $line);
       foreach my $word (@words) {
           #print "Added: $word\n";
           push(@dalechall, $word);
       }
  }
  return @dalechall;
}

  my $fleschIndex=buildFleschObject($ARGV[0]);
  my $a = ($fleschIndex->{syllableCount})/($fleschIndex->{wordCount});   #
  my $b = ($fleschIndex->{wordCount})/($fleschIndex->{sentenceCount});  #Variables used for respective equations
  my $aD = ($fleschIndex->{diffWordCount})/($fleschIndex->{wordCount}); #
  print("Difficult Word: $fleschIndex->{diffWordCount}\n");
  
  my $readIndex = 206.835-($a*84.6)-($b*1.015);        #Flesch-Readability Index Variable
  my $gradeIndex = ($a*11.8)+($b*0.39)-15.59;          #Flesch Grade Level Index
  my $dalePercent = ($aD * 100);                      #Dale-Chall Percentage
  
  my $daleIndex =($dalePercent*0.1579)+($b*0.0496);    #Dale-Chall Raw Score
  if($dalePercent > 5)
  {
    $daleIndex += 3.6365;
  }
  print "Sentence Count: $fleschIndex->{sentenceCount}\n";
  print "Word Count: $fleschIndex->{wordCount}\n";
  print("Syllable Count: $fleschIndex->{syllableCount}\n");
  print("\nReadability and Grade Index\n");
  print("-----------------------------------\n");
  print("Flesch Readability Level: $readIndex\n");
  print("Flesch Grade Level Index: $gradeIndex\n");
  print("Dale-Chall Readability Level: $daleIndex\n");